[MusicDB] Add artist roles to handle Composer, Conductor, DJMixer etc. Tags

DaveTBlake opened the pull request:
First step in improving browsing for classical music is to process the standard tags for composer (TCOM), conductor (TPE3) and Ensemble.  This could be considered a work in progress, but could be merged as an initial phase providing limited user improvements.

Unlike popular music albums which usually have one album artist, classical music frequently has multiple artists - soloists, conductor, orchestra and composer. The composer is the primary element used when selecting what you want to play, and although not a performer most classical music files will have composer name in the artist (or album artist) tag just to make them accessible on the various players (inc. Kodi).  Musicbrainz also treats composer as an artist.

Once the artist list includes a mix of composers, orchestras, conductors, soloists (as well as pop artists)  it is useful to be able to see just one kind of artist at a time. Lack of player support for these tags means that users will have incorporated them in the artist and album artists tags. Rather than have extra tables or song attributes for these tags, my idea is to use these tags to identify the "role" that an artist has in the recording. An approach that could easily be extened to other as yet unsupported tags, but wanting to try the concept first. This means a database change adding "role" as a attribute of the song_artist and album_artist tables.

The UI aspect of this change is made available through folder type custom nodes. For example

```
<?xml version='1.0' encoding='UTF-8'?>
<node order="1" type="folder">
    <label>Composers</label>
    <path>musicdb://genres/-1/?roleid=3</path>
</node>
```

Implementing this change I have looked at the inital work on artist credits discussion between Night199UK and JMarshall. The original intentions to use JoinPhrase to construct artist names and get rid of the various strArtist and use of strutils::Join everywhere. This did not happen, and despite being used to store " / " JoinPhrase is not used, neither is bFeatured. I have therefore removed these attributes (replacing them with role). I really have checked carefully before doing that.

Not sure how much more of the design I should describe at this point, or if there is somewhere better for design documentation? I am hoping we can move the music library forward despite the lack of experienced dev team members even using the music side let along having time to consider design changes.

From other PR it seems that @Montellese, @evilhamster, @Razzeee, and @zag2me are most likey to have an interest in music library/tags. Lots to discuss I am sure!!

Razzeee commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -981,27 +999,16 @@ bool CMusicDatabase::GetAlbum(int idAlbum, CAlbum& album, bool getSongs /* = tru
 
+    //Get album, song and album song info data using separate queries/datasets because we can have 
+    //multiple roles per artist for both album and songs and that makes a single combined join impractical
+    //Get album datae
```
typo

Razzeee commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -981,27 +999,16 @@ bool CMusicDatabase::GetAlbum(int idAlbum, CAlbum& album, bool getSongs /* = tru
+    sql = PrepareSQL("SELECT albumview.*,albumartistview.* "
+                     " FROM albumview "
+                     " LEFT JOIN albumartistview ON albumview.idAlbum = albumartistview.idAlbum "
+                     " WHERE albumview.idAlbum = %ld "
```
it doesn't seem right to use %ld here isn't %o correct? Not sure if our PrepareSQL can handle that, just saying.

DaveTBlake replied:
Just being consistent with existing code, so PrepareSQL must be handling it I guess. Should all the %ld be %o?

Razzeee replied:
Still learning C++, but my research this afternoon said %o. I may be wrong.

Razzeee commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -981,27 +999,16 @@ bool CMusicDatabase::GetAlbum(int idAlbum, CAlbum& album, bool getSongs /* = tru
-
+    sql = PrepareSQL("SELECT albumview.*,albumartistview.* "
+                     " FROM albumview "
+                     " LEFT JOIN albumartistview ON albumview.idAlbum = albumartistview.idAlbum "
```
I tink a simple join should be okay here, did a quick test without any problems. Will also speed things up a little.

DaveTBlake replied:
It depends if we can have albums without album artists? The original Kodi processing (pre Musicbrainz) will ensure that every album has at least one album artist, if it is varied then it is set to "various artists" etc. However if the album has MBID then this is not enforced. Although Picard tags files with an album artist(s), it is possible for users to mess with this tagging (something that happens even if we don't like it) and remove the album artist. This means we could have albums without artists in real life user music libaries. An outer join here means that we would still see these (slightly messed up) albums.
So what do we do anyone?

DaveTBlake replied:
Having checked the history JMarshall (who seems to have written much of this part) deliberately changed these to LEFT JOIN because albums and songs can exist without artists. See https://github.com/xbmc/xbmc/commit/70ee5f75a4d622757cb473fbb89c414ba1349592

Razzeee replied:
Then keep it with left joins, the performance decrease is not that big with such a simple query.

Razzeee commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1011,60 +1018,102 @@ bool CMusicDatabase::GetAlbum(int idAlbum, CAlbum& album, bool getSongs /* = tru
+    {
+      sql = PrepareSQL("SELECT songview.*, songartistview.*"
+        " FROM songview "
+        " LEFT JOIN songartistview ON songview.idSong = songartistview.idSong "
```
Don't use left joins when you can use a inner join

DaveTBlake replied:
OK. But are we absolutely sure there are no songs withou song artists?

Tolriq replied:
I can confirm for sure that there's case where albums / songs have no artists it's for singles as I use that to detect those singles since there's other problems with some returned fields to JSON.

Razzeee replied:
Then my sample database with 22.000 songs doesn't hit this case. Will try tomorrow with another database that I know has single songs.

zag2me commented:
Excellent, thanks for getting stuck into this @DaveTBlake. 

The design and concept is good, and I like the way you are thinking about the different cases for tagging. The new "Role" DB field is an great idea and could work with electronic artists as well. I think in general we should follow how musicbrainz classifies artists.

As always, can't comment too much on the code but the change in general looks good.

One small thing is that maybe its a good idea to add new tags in separate commits.

evilhamster commented:
You should remove the two commits "Change the way the (album)artists tag are handled (#7486 not yet merged" and "Various artists" album artist in the artists list (#7938 not yet merged)" from the PR since one of them have already been merged and the other one is handled in an other PR.

For readability it might be a good idea to split into several commits (maybe one for removing strArtist /JoinPhrase and one for adding roles)?

As we discussed over PM's, I would prefer if the role are not defined by enum's but rather as strings (the idRole could reference a table with the different roles description). There are some use cases where this would be nneded; TMCL tags would be one, being able to add relationships from musicbrainz would be an other. 

Haven't been able to read all the code yet, but started with the tagloader lib parts and it looks like you only load the first entry of a tag even though they should be able to contain a list of items.

Also it looks like you are changing the upgrade steps for db version 54, isn't it a better idea to bump db version to 55 so you can separate your changes (and make sure they execute for the users that are on 54)?

Will try to check more later.

jjd-uk commented:
@evilhamster is your provided forum email address up to date, something was sent to you on behalf on the Team and just wanted to check if you got it.

Razzeee commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -3925,8 +3966,29 @@ void CMusicDatabase::UpdateTables(int version)
+    m_pDS->exec("DROP TABLE album");
+    m_pDS->exec("ALTER TABLE album_new RENAME TO album");
+
+    //Remove strJoinPhrase, boolFeatured from song_artist table and add idRole
```
as  @evilhamster pointed out this part below needs to be in a new db schema version

Razzeee commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -3867,15 +3908,15 @@ void CMusicDatabase::UpdateTables(int version)
-      //Remove dateAdded from artist table
-      m_pDS->exec("CREATE TABLE artist_new ( idArtist integer primary key, "
+    //Remove dateAdded from artist table
+    m_pDS->exec("CREATE TABLE artist_new ( idArtist integer primary key, "
```
Are you changing the indentation by accident here?

DaveTBlake replied:
Possibly getting confused with the 2 space or 4 space tabs on different files. Trying to stick to recommened 2 space, but also be consistent with the exiting code if it is 4 space. Will review.

DaveTBlake commented:
Thanks for checking this out. Will make code corrections as suggested, keep them comming!

@evilhamster 
As for the first 2 commits covered by other PR, I just don't know how else to handle other changes to the same piece of code. As it is this will have to be manually merged anyway, and other music library changes are bound to overtake it too (pulled, built and merged while we are still considering this). I don't want to remove those commits because I need the code to be like they leave it.

I understand that smaller commits are easier to review, but again I don't see how to get there from here. Adding roles without removing JoinPhrase I would have had to add processing for JoinPhrase, removeal made the most sense. This is as small a commit I could create and still have a working result that did something.

Again database version changed like quick sand under my feet!! Did my best to keep aligned, probably still made an inconsistent result. Github novice struggling to get it right!

Montellese commented:
Having commits from other open PRs is ok if the changed depend on them. Once they have been merged the branch can be rebased and the commits will (more or less) automatically disappear. We just need to remember that this can only be merged after the other PRs.

DaveTBlake commented:
Thanks @Montellese for clarifying. I will leave the first 2 commits as a clear reminder of the merge dependency.

As this work becomes increasing out of date with other smaller music changes being merged in the meantime (it feels like a continuous stream at the moment), is there anything I can do that will help with the review process and the (hopefully) subsequent merge? For example if a give my branch a new DB version, as Evilhamster and Razzee point out, do I also bring it inline with the current master, or do I pretend I don't know there are other changes? Advice appreciated.

Razzeee commented:
@DaveTBlake 
You will need to rebase this so that it is mergeable again.
Are you using any tool for git or git itself? If I may ask?

That will also pull in this work: https://github.com/xbmc/xbmc/pull/8012 you should then have a look at the code again, as your code should be in the same format.

DaveTBlake commented:
@Razzeee 
I have installed TortoiseGit, but might be better just using command line. Handholding always appreciated :)

At what point do I rebase? Music changes seem to be comming thick and fast at the moment, so do I do it now or wait for more comments (and commits) I know that you and evilhamster have only just started looking at this.

I spotted #8012, but sure other music work could get past me.

Razzeee commented:
You might want to backup your folder the first few times you do this.
Rebases will probably be easier if you do them as often as possible and with as few commits as possible.

Just choose 'Rebase...' from the context menu on your folder. In that dialog, be sure that you are on this branch. If you don't have the official kodi repo as remote configured hit '...' and add a new remote with the url https://github.com/xbmc/xbmc.git, be sure to fetch it. Then you can select whatever-you-named-it/master as Upstream.
You should now see your three commits, we're basically getting a fresh master and adding your three commit ontop of it. Hit 'Start rebase', if you hit conflicts, resolve them and tell tortoise to move on (same button position as start) and if your done, check the log and try to compile. Just so we're sure everything is fine.
If you have done that and everything is fine, just push and select the checkbox for 'unknown changes'. Double check your remote, it should be your remote NOT the kodi remote. (you shouldn't have rights but it's always good to know where your pushing to)

DaveTBlake commented:
@evilhamster
>... the tagloader lib parts and it looks like you only load the first entry of a tag even though they should be able to contain a list of items.

A song will definitely only have one conductor, any extra tag is a mistake and reasonable to ignore. I was going to say that a song only has one composer, but now I think about it whilst this is true for many classical works it is not globally true. Lenon & Macartney spring to mind (not that popular music listeners usually bother with composer, they might start), or Gilbert & Sulliven for opereta, so you have a point. But is TCMP multivalue?? Have to check, FlAC/Vorbis/Xiph will cope.
I guess ensemble could be multiple, if you lump orchestra and choir together for say Beethoven's 9th, have to find some examples.

It could make to a long track description to display during play, but good catch.

evilhamster commented:
@jjd-uk 
Yupp, I got it but somehow I manage to miss to send a reply, thanks for  reminding me!

@DaveTBlake 
I think they are rare, but there are cases where performances have used more than one conductor (the one I read about was https://en.wikipedia.org/wiki/Gruppen, but there are probably more).

One thing I have been thinking about, even though TPE1/TPE2 can be multi-value it's mostly used as a single string that reflects what is written on the album, do you think it could be the same problem with these tags?

tcmp is itunes compilation thingy, so I hope that one isn't multivalue :)

In std::string GetArtistString, could there be a problem with duplicate names in artistString if credits contain the same artist both as albumartist and artist for a song? 

Btw, I have been thinking some more about using one additional table versus using enum definitions and I'm not completely convinced but and I'm starting to think you made the correct call in this pr (so you can disregard my comment about it in my first comment).

But it's great work, let me know if you need help with the rebasing of it

DaveTBlake commented:
Rebased so now upto date with master (thanks guys for the hand holding). A build could help with getting some others testing with appropriately tagged classical music. How does that happen?

@evilhamster 
I am thinking that starting simple with one composer, conductor and ensemble could be the way to go. It is a huge leap forward,and handling all multi-line possibilities could just complicate things at this point for little return.

>One thing I have been thinking about, even though TPE1/TPE2 can be multi-value it's mostly used as a single string that reflects what is written on the album, do you think it could be the same problem with these tags?

Yes, almost certainly. It is down to the tagging software too. FLAC/Vorbis/Xiph comments for ARTIST can be multiple but usually you get one with separators for the many names.

TCMP was a typo, I meant TCOM of course!

>In std::string GetArtistString, could there be a problem with duplicate names in artistString if credits contain the same artist both as albumartist and artist for a song? 

I don't think so, different tables are involved (in that sense the separate roles are unnecessary, but I feel they help with debugging if nothing else). 

However I do wonder about how well Kodi keeps the 3 ways it holds artist names in line - string, vector and artist credit objects - even without this PR. The string (and denormalisation when stored in database) does make sense as a display value, the artist credits object makes sense too. Then there is the vector, which is historic and gets passed about lots. I do wonder if there is an arguement for rationalisation or at least replacing the vector with a method.

zag2me commented:
![icone_windows](https://cloud.githubusercontent.com/assets/1050512/9930234/2b79c1c6-5d2a-11e5-831b-21d5a0785bb8.png) Windows test build
http://mirrors.kodi.tv/test-builds/win32/KodiSetup-20150916-0fe4fef-ArtistRoles.exe

Razzeee commented:
Whats the status here? Has there been any testing? Any more work you want to do?
This also needs a rebase.

DaveTBlake commented:
@Razzeee it has received some testing and been OK, but probably needs a second team member to sign off on it. Happy to rebase, but I feel it would be better to get #8158 done first so my focus is on that, and that could provoke some modifications.

Meanwhile there is a build that can be used for testing further, just need some classical music users to engage with suitably tagged music (otherwise you don't see any difference).

DaveTBlake commented:
Have rebased just to show proof of life!! I would like to rebase again to include #8158 (as yet unmerged) so will probably keep that on my repo to avoid confusion. Either that or wait... :(

Razzeee commented:
Can you rebase and squash it to one commit?

DaveTBlake committed:
Handling of tags giving people that contribute to the recording e.g.
Composer, Conductor, Lyricist, Mixer etc. including TIPL and TMCL held as
vector of artist name and role in CMusicInfoTag and CSong. Contributors
are held in library as artists with role in song_artist and role tables determining which tag was source of data.

Remove JoinPhrase and boolFeatured as unused elsewhere, artist string now stored rather than built.

Queries changed to generally access only standard artists not other roles

Contributors included in both JSON and Info Labels

Add example/default role nodes and modified SongInfoDialog so data
visible.

DaveTBlake commented:
Tried to rebase again, hard to keep up! Manually merging was painful and I ended up creating a new branch on my repo (with an uptodate master). Working through manually reapplying changes led me to modify the design a little for the better, so worthwhile. The new, mergeable, version of this change is in a new branch on my repo https://github.com/DaveTBlake/xbmc/tree/ClassicalContributors, just not sure how to push it here?

Razzeee commented:
Make a backup.
Switch to this branch.
Hard reset it to the last version without any of your changes.
Rebase against your new branch. (you could also Cherrypick the commits)

Might work, haven't tried obviously. Using any git gui or just command shell?

DaveTBlake commented:
@Razzeee my hard reset seems to have closed this PR. Anyway to reopen it? My associated branch is ready now, or do I open another PR?

DaveTBlake commented:
@Razzeee OK so I seem to have managed to reopen this, and it has my recent work based on a recent head , but it says that there are conflcits to be resolved. I am beginning to loose the will to live!

Not sure where the issue is, but I know there are lots of changes to musicdatabase.cpp that is why I did a rework. What now (please don't say rebase again)?

phate89 commented:
You have to rebase because there are conflicts. The code is different than the one you edited for the price and github doesn't know what's the right one.
I'm not a git expert so I saved this stackoverflow discussion to know what to do:
http://stackoverflow.com/questions/7244321/how-to-update-a-github-forked-repository
Of course instead of 'git checkout master' you need 'git checkout ArtistRoles' and instead of 'git push -f origin master' you need 'git push -f origin ArtistRoles'. Of course since there are conflicts when you do 'git rebase upstream/master' will prompt you that can't rebase because there are conflicts in some files and it will give you a list of those files. Then you need to manually open each file and fix them. One of this will be for sure 'xbmc/music/MusicDatabase.cpp.. in those files you will find some piece of code like this:
<<<<<<<<HEAD
here's the code from head
==========
Here's the code from your pr
>>>>>>>>name of your pr
After you fixed ALL these changes in every file with conflicts you will need to add them back with 'git add', for example:
git add xbmc/music/MusicDatabase.cpp
After you added all the files you continue the rebase with 'git rebase --continue' and it will finish correctly. After that you force push to your branch with 'git push -f origin ArtistRoles'

DaveTBlake commented:
Phew! Fun with Git on a Friday night, thanks for the help @phate89. Rebased (again) and up-to-date.
 @evilhamster  @Razzeee, @Montellese  can we look to merging this before I have to go though that again.

hudokkow commented:
So much git-fu fun... :wink: 
jenkins build this please

Razzeee commented:
Please try to compile it again, there seems to be a problem, that causes all platforms to fail in jenkins.

hudokkow commented:
Well, just checked linux64 and win32 build logs. Error is very explicit:

<code>Song.cpp:79:3: error: 'Year' was not declared in this scope</code>

DaveTBlake commented:
A merge/Git mistake on my part. Not having a good evening :(

hudokkow commented:
Don't worry about it. Happens to us all. :wink: 
jenkins build this please and behave this time

DaveTBlake commented:
Ok have gone for another rebase just to try and keep current. Not wanting a merge any time soon because I would like adjust a few bits to work better. Nothing for anyone to review just yet, been distracted by other fixes!

DaveTBlake commented:
I have reworked this PR incorporating lots of good suggestions from @evilhamster  and broadened the scope so offering general improvement not just classical music specific.

It adds processing for the following tags:
COMPOSER (TCOM)
CONDUCTOR (TPE3)
LYRICIST (TEXT)
REMIXER (TPE4)
ENSEMBLE / BAND / ORCHESTRA (TXXX:Ensemble)
ARRANGER,
ENGINEER,
PRODUCER,
DJMIXER,
MIXER
PERFORMER=artist (instrument) (TIPL and TMCL)

These tags are all about people (or groups of people) that contribute in some way to the recording as musicians or are in some other way involved. They are added to the music library as artists with a role. They can be listed separately from general artists (those scanned from the albumartist and artist tags) using custom nodes which can be configured to reflect the users interests. They will not be displayed on the original artist lists, so those not interested in this extra data will be uneffected. 

I have modified JSON interface to return the stored data for songs, and added Info lables (no point having the data unless it is made available. Changes to the Song Info Dialog are needed to make the additional data visible.

A test build is already available http://mirrors.kodi.tv/test-builds/win32...People.exe
You will also want to download the following custom node files https://www.dropbox.com/s/6jhkz3chn2vf25...s.zip?dl=0  

User testing is being discussed here http://forum.kodi.tv/showthread.php?tid=213153&page=7
I would like to get this merged because I am fed up with rebasing, but realise it is a bigger step in one PR than I originally intended.

Paxxi commented on xbmc/GUIInfoManager.cpp:
```diff
@@ -3373,6 +3386,30 @@ bool CGUIInfoManager::GetMultiInfoBool(const GUIInfo &info, int contextWindow, c
+        std::string strRole = m_stringParameters[info.GetData1()];
+        // Find value for role if already stored
+        int artistcount = -1;
+        for (std::vector<std::pair<std::string, int>>::iterator i = m_libraryRoleCounts.begin(); i != m_libraryRoleCounts.end(); ++i)
```
needs {} for readability, someone editing this will break it
could also use a for (const auto & role : m_libraryRoleCounts) for readability or at least auto i = m_libraryRoleCounts.begin()

Paxxi commented on xbmc/music/Album.cpp:
```diff
@@ -99,11 +98,11 @@ CAlbum::CAlbum(const CFileItem& item)
-      std::string strJoinPhrase = (it == --tag.GetAlbumArtist().end() ? "" : g_advancedSettings.m_musicItemSeparator);
-      CArtistCredit artistCredit(*it, "", strJoinPhrase);
+      CArtistCredit artistCredit(*it);
       artistCredits.push_back(artistCredit);
```
merge these two lines and use artistCredits.emplace_back(*it); for some "bonus performance", it avoids some copying.

Paxxi commented on xbmc/music/Album.cpp:
```diff
@@ -200,9 +199,12 @@ const std::string CAlbum::GetAlbumArtistString() const
   //but is takes precidence as a string because artistcredits is not always filled during processing
   if (!strArtistDesc.empty())
     return strArtistDesc;
+  std::vector<std::string> artistvector;
```
std::vector<std::string> artistvector(artistCredits.size()); to set the size directly as we know it

Paxxi commented on xbmc/music/Album.cpp:
```diff
@@ -200,9 +199,12 @@ const std::string CAlbum::GetAlbumArtistString() const
     return strArtistDesc;
+  std::vector<std::string> artistvector;
+  for (VECARTISTCREDITS::const_iterator i = artistCredits.begin(); i != artistCredits.end(); ++i)
+    artistvector.push_back(i->GetArtist());
```
same here, can use emplace_back for some bonus perf, not that we'll notice it but it feels nice

Paxxi commented on xbmc/music/Artist.h:
```diff
@@ -130,21 +132,49 @@ class CArtistCredit
+public:
+  CMusicRole() { }
+  CMusicRole(std::string strRole, std::string strArtist) : idRole(-1), m_strRole(strRole), m_strArtist(strArtist), idArtist(-1) { }
+  CMusicRole(int role, std::string strRole, std::string strArtist, long ArtistId) : idRole(role), m_strRole(strRole), m_strArtist(strArtist), idArtist(ArtistId) { }
```
don't use long, prefer int64_t or uint64_t if you want a 64-bit type or int/int32_t/uint32_t for 32 bit types

DaveTBlake replied:
Going to stay with long (for now) to match how artistId is treated in the rest of artist.h, however I think we need to review the type for artistId and albumId. It is long in some places, but most of musicdatabase.cpp processes them as int. Probably needs to be int64_t everywhere.

Tolriq commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -416,7 +416,8 @@
       "enum": [ "title", "artist", "albumartist", "genre", "year",
-                "rating", "votes", "userrating", "album", "track", "duration", "comment",
+                "rating", "votes", "userrating", "album", "track", "duration", "comment", 
+				"mood", "contributors", "composer", "conductor", "orchestra", "lyricist",
```
To keep backward compat with non named params, new fields needs to always be added at the end.

+ Some indentation problem

Tolriq commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -495,6 +507,11 @@
+	  "composer": { "type": "string"},
+	  "conductor": { "type": "string"},
+	  "orchestra": { "type": "string"},
+      "lyricist": { "type": "string"},
```
To keep backward compat with non named params, new fields needs to always be added at the end.

+ Some indentation problem

Tolriq commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -429,6 +430,17 @@
+      "properties": {
+        "name": { "type": "string", "required": true },
+        "role": { "type": "string", "required": true },
+		"artistid": { "type": "integer", "required": true }
```
Indent

Paxxi commented on xbmc/music/Artist.h:
```diff
@@ -130,21 +132,49 @@ class CArtistCredit
+  std::string GetRoleDesc() const { return m_strRole; }
+  int GetRoleId() const { return idRole; }
+  long GetArtistId() const { return idArtist; }
+  void SetArtistId(const long iArtistId) { idArtist = iArtistId;  }
```
remove const, it's passed by value so it doesn't do anything other than cause a warning that you're assigning from const to non-const variable

Paxxi commented on xbmc/music/Artist.h:
```diff
@@ -130,21 +132,49 @@ class CArtistCredit
+  long idArtist;
+};
+
+typedef std::vector<CMusicRole> VECMUSICROLES;
```
not really needed, I personally don't like these typedefs but some disagree so you can decide yourself what's best :)

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -593,11 +604,15 @@ bool CMusicDatabase::UpdateAlbum(CAlbum& album, bool OverrideTagData /* = true*/
           std::distance(song->artistCredits.begin(), artistCredit));
       }
+      // Having added artist credits (maybe with MBID) add the other contributing artists (MBID unknown)
+      for (VECMUSICROLES::const_iterator credit = song->GetContributors().begin(); credit != song->GetContributors().end(); ++credit)
```
you're looping over these in two places, might be better to add an overloaded AddSongContributor that takes a vector and hide the looping in there

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1355,26 +1378,125 @@ bool CMusicDatabase::DeleteArtistDiscography(int idArtist)
+    strSQL = PrepareSQL("SELECT idRole FROM role WHERE strRole LIKE '%s'", strRole.c_str());
+    m_pDS->query(strSQL);
+    if (m_pDS->num_rows() > 0)
+      idRole = (int)m_pDS->fv("idRole").get_asInt();
```
no need for the cast, get_asInt() returns int

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1355,26 +1378,125 @@ bool CMusicDatabase::DeleteArtistDiscography(int idArtist)
+    {
+      strSQL = PrepareSQL("INSERT INTO role (strRole) VALUES ('%s')", strRole.c_str());
+      m_pDS->exec(strSQL);
+      idRole = (int)m_pDS->lastinsertid();
```
this could result in data loss, not sure if we'll hit 2^31 roles though
if it remains it should be static_cast<int>(m_pDS->lastinsertid());

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -1355,26 +1378,125 @@ bool CMusicDatabase::DeleteArtistDiscography(int idArtist)
+    strSQL = PrepareSQL("SELECT idArtist FROM song_artist WHERE idSong = %i AND strArtist LIKE '%s' ", idSong, strArtist.c_str());
+    m_pDS->query(strSQL);
+    if (m_pDS->num_rows() > 0)
+      idArtist = (int)m_pDS->fv("idArtist").get_asInt();
```
cast not needed

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -3282,6 +3433,67 @@ bool CMusicDatabase::GetYearsNav(const std::string& strBaseDir, CFileItemList& i
+    if (!m_pDS->query(strSQL))
+      return false;
+    int iRowsFound = m_pDS->num_rows();
+    if (iRowsFound == 0)
```
no need to create a variable here, can just call num_rows() in the if as the value isn't used anywhere else

DaveTBlake replied:
But it sure does help with debugging. For that reason, and because there are many other places in this unit that do the same I am going to keep it as is if that is OK.

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -3282,6 +3433,67 @@ bool CMusicDatabase::GetYearsNav(const std::string& strBaseDir, CFileItemList& i
+    while (!m_pDS->eof())
+    {
+      std::string labelValue = m_pDS->fv(1).get_asString();
+      CFileItemPtr pItem(new CFileItem(labelValue));
```
use auto pItem = std::make_shared<CFileItem>(labelValue);

it's more general and easier to understand what it does, especially for people coming from other c++ code bases

DaveTBlake replied:
Unfortunately make_shared doesn't have a version for string argument, or have I misunderstood?

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -5719,6 +5964,23 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
+    idRole = 1;  //Restrict to song_artist to "artists" only, no other roles.
+  
+  std::string strRoleSQL; //Role < -1 means all roles, otherwise filter by role
+  if(idRole > 0) strRoleSQL = PrepareSQL(" AND song_artist.idRole = %i ", idRole);  
```
can idRole ever be 0? wondering because it's either 1 or something larger I assume reading the flow.

DaveTBlake replied:
Read in as a custom node parameter idRole could be any int.  Following other parameters -1 means not specified, and although there isn't a 0 role it could be specified and return no records, just like artistId or genreId etc. Because of the nature of role I am adding the option that a value < -1 means give me all roles.

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -5719,6 +5964,23 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
+    if (option != options.end())
+      idRole = GetRoleByName(option->second.asString());
+  }
+  if (idRole == -1)
```
why initialize to -1 and then set it to 1 here? Could just init it to 1

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -5760,30 +6022,59 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
+      if (!albumArtistsOnly || idRole > 1)  //Song and album artists or album artists with specific role, check song_artist
+        strSQL += PrepareSQL("(SELECT song_artist.idArtist FROM song_artist WHERE song_artist.idRole = %i)", idRole);
+
+      if (idRole < -1) //All artists contributing to songs, all roles.
```
should it really be < -1? I haven't seen it set to lower than -1 so maybe <= or == ?

DaveTBlake replied:
Yes really <-1, see comment above

Paxxi commented on xbmc/music/MusicDatabase.cpp:
```diff
@@ -5819,32 +6110,26 @@ bool CMusicDatabase::GetFilter(CDbUrl &musicUrl, Filter &filter, SortDescription
     if (option != options.end())
     {
-      int idArtist = static_cast<int>(option->second.asInteger());
+      int idArtist = (int)option->second.asInteger();
```
talking about auto, it could be used here since we explicitly cast the value to int, it's clear that it's an int. Might be a bad example as auto is more typing but anyway, just thought I'd give an example

Paxxi commented on xbmc/music/MusicDatabase.h:
```diff
@@ -340,6 +345,9 @@ class CMusicDatabase : public CDatabase
 
   int GetSinglesCount();
+
+  int GetArtistCountForRole(int Role);
```
lower case first letter for variables

Paxxi commented on xbmc/music/tags/MusicInfoTag.cpp:
```diff
@@ -756,6 +761,19 @@ void CMusicInfoTag::Serialize(CVariant& value) const
   value["musicbrainztrmid"] = m_strMusicBrainzTRMID;
   value["comment"] = m_strComment;
+  value["contributors"] = CVariant(CVariant::VariantTypeArray);
+  for (unsigned int i = 0; i < m_musicRoles.size(); ++i)
```
use for (const auto& role : m_musicRoles and then access as role.GetArtist(); instead of m_musicRoles{i].GetArtist();

Paxxi commented on xbmc/music/tags/MusicInfoTag.cpp:
```diff
@@ -830,6 +848,14 @@ void CMusicInfoTag::Archive(CArchive& ar)
     ar << m_lastPlayed;
     ar << m_dateAdded;
     ar << m_strComment;
+    ar << (int)m_musicRoles.size();   
```
this can result in data-loss, .size() returns uint64_t on some platforms, I doubt we'll run into the issue though

Paxxi commented on xbmc/music/tags/MusicInfoTag.cpp:
```diff
@@ -867,6 +893,22 @@ void CMusicInfoTag::Archive(CArchive& ar)
+      ar >> strArtist;
+      ar >> idArtist;
+      CMusicRole Credit(idRole, strRole, strArtist, idArtist);
+      m_musicRoles.push_back(Credit);
```
can be called as m_musicRoles.emplace_back(idRole, strRole, strArtist, idArtist); to avoid copying the item.

Paxxi commented on xbmc/music/tags/MusicInfoTag.cpp:
```diff
@@ -956,6 +999,64 @@ void CMusicInfoTag::AppendGenre(const std::string &genre)
+const std::string CMusicInfoTag::GetArtistStringForRole(const std::string& strRole) const
+{
+  std::vector<std::string> artistvector;
+  for (VECMUSICROLES::const_iterator credit = m_musicRoles.begin(); credit != m_musicRoles.end(); ++credit)
```
add {} for readability

Paxxi commented on xbmc/music/tags/MusicInfoTag.cpp:
```diff
@@ -956,6 +999,64 @@ void CMusicInfoTag::AppendGenre(const std::string &genre)
+  std::string strLabel;
+  for (VECMUSICROLES::const_iterator credit = m_musicRoles.begin(); credit != m_musicRoles.end(); ++credit)
+  {
+    std::string character;
```
can skip this and just strLabel += in both cases

Paxxi commented on xbmc/music/tags/MusicInfoTag.h:
```diff
@@ -154,12 +155,22 @@ class CMusicInfoTag : public IArchivable, public ISerializable, public ISortable
+  void AddArtistRole(const std::string& Role, const std::vector<std::string>& artists);
+  void AppendArtistRole(const CMusicRole& ArtistRole);
+  const std::string GetArtistStringForRole(const std::string& strRole) const;
+  const std::string GetContributorsText(bool bIncludeRole = false) const;
```
avoid using bool parameters, it's very hard to know what they do without inspecting the method declaration. I would go with GetContributorsText() and GetContributorsAndRolesText() or similar

Paxxi commented on xbmc/music/tags/TagLoaderTagLib.cpp:
```diff
@@ -302,6 +326,27 @@ bool CTagLoaderTagLib::ParseTag(ID3v2::Tag *id3v2, MUSIC_INFO::EmbeddedArt *art,
+      // In fieldlist every odd field is a function, and every even is an artist or a comma delimited list of artists.
+      for (ID3v2::FrameList::ConstIterator ip = it->second.begin(); ip != it->second.end(); ++ip)
+      {
+        ID3v2::TextIdentificationFrame *tiplframe = reinterpret_cast<ID3v2::TextIdentificationFrame*> (*ip);
```
you want dynamic_cast < ID3V2::TextIdentificationFrame* > (*ip) here, a reinterpret_cast will just cast it and funny things might happen

Paxxi commented on xbmc/music/tags/TagLoaderTagLib.cpp:
```diff
@@ -302,6 +326,27 @@ bool CTagLoaderTagLib::ParseTag(ID3v2::Tag *id3v2, MUSIC_INFO::EmbeddedArt *art,
+      for (ID3v2::FrameList::ConstIterator ip = it->second.begin(); ip != it->second.end(); ++ip)
+      {
+        ID3v2::TextIdentificationFrame *tiplframe = reinterpret_cast<ID3v2::TextIdentificationFrame*> (*ip);
+        if (!tiplframe) continue;
```
change this to if (tiplframe) and get rid of the continue, it's not needed for such a small loop

Paxxi commented on xbmc/music/tags/TagLoaderTagLib.cpp:
```diff
@@ -302,6 +326,27 @@ bool CTagLoaderTagLib::ParseTag(ID3v2::Tag *id3v2, MUSIC_INFO::EmbeddedArt *art,
+        if (!tiplframe) continue;
+        AddArtistRole(tag, StringListToVectorString(tiplframe->fieldList()));
+      }
+    else if (it->first == "TMCL")
```
braces for readability

DaveTBlake replied:
Beg to differ in the context, the pattern in that function is more readable without

DaveTBlake commented:
Thanks @Paxxi for the code improvements, and @Tolriq for the JSON suggestions. 

I have modified the JSON interface to avoid parameter by position compatibility conflicts (including fixing those introduced by #8405), added role as  filter parameter to GetArtists (so it has the same flexibile filtering as Kodi itself) and contributors etc. fields to Player.GetItem to retrieve the new data for the currently playing song. There may be other enhancements to be made in a separate PR, but I would like to get this in place as a good starting point.

Actually I would really like to get this merged early in the cycle, ASAP if no-one minds. The rebasing is getting tedious, and once part of the nightly builds it will receive better testing. Then I can build on what is here with subsequent changes, rather than it just grow.

@Montellese can you give the JSON stuff your approveal?
@zag has taken and early build it for a test drive along  with some other users and nothing broke!

Once this is out I am hoping that @HitcherUK will come up with a Song Info diaglog that shows the additional data, and looking to.

Tolriq commented on xbmc/interfaces/json-rpc/schema/methods.json:
```diff
@@ -641,6 +641,8 @@
           { "type": "object", "properties": { "albumid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false },
           { "type": "object", "properties": { "album": { "type": "string", "minLength": 1, "required": true } }, "additionalProperties": false },
           { "type": "object", "properties": { "songid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false },
+		  { "type": "object", "properties": { "roleid": { "$ref": "Library.Id", "required": true } }, "additionalProperties": false },
```
Indent.

Tolriq commented:
@DaveTBlake let's ignore the contributor flag, but adding filtering by roleid without any way to get or guess the value should be avoided.

Filter by roles with no GetRoles is limited but why not.

But GetArtist description should be updated to explains how this works because currently the filters works the way around if I follow our discussion.

Meaning by default contributors are not returned but if you filter them then they are returned. 
All other filters works the opposite : return everything, if you apply filter then you get less.

If by default GetArtists return contributors only artists then it's a deal breaker without a flag.

DaveTBlake commented:
Jenkins build this please

DaveTBlake commented:
@Tolriq thanks for your helpful comments about JSON API. I have added GetContributorRoles method and role as a filter parameter to GetSongs.  I agree that the role criteria use needs description, but not sure where to put it?

When it comes AudioLibrary.GetArtists returning denormalised data, such as flagging the contributor only artists, or even the roles that artist plays, in the same way as you want it to return a compilationartist flag, that seems are far bigger issue than the scope of this PR. Currently the "compilationartist" field is always false, fixing that and returning roles too is something for another PR.

DaveTBlake commented:
@Montellese your overview of the JSON additions would be appreciated.

Montellese commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -416,7 +428,8 @@
       "enum": [ "title", "artist", "albumartist", "genre", "year",
-                "rating", "votes", "userrating", "album", "track", "duration", "comment",
+                "rating", "votes", "userrating", "album", "track", "duration", "comment", 
+                "mood", "contributors", "displaycomposer", "displayconductor", "displayorchestra", "displaylyricist",
```
The new field values should go to the end.

Montellese commented:
JSON-RPC changes look ok apart from the minor. You're also missing the JSON-RPC API bump.

DaveTBlake commented:
Thanks @Montellese will fix enum order (and those added by #8405 in the middle too). Must have lost the API bump in the rebase, I remember doing it.

Tolriq commented on xbmc/interfaces/json-rpc/schema/types.json:
```diff
@@ -390,6 +390,18 @@
+    "properties": {
+      "roleid": { "$ref": "Library.Id", "required": true },
+      "title": { "type": "string" },
+      "thumbnail": { "type": "string" }
```
Is there really thumbnails for roles or a too fast copy paste from genre ?

DaveTBlake replied:
A silly. I had a syntax bug, did some re-copying to try and find it and the thumbnails sneeked back in. Will fix of course.

Tolriq commented:
Just so that there's a public record about it :)

I do not agree on usability of current JSON API exposition of this, to be really useful for remote makers :)

DaveTBlake commented:
Have fixed enum order, bumped JSON. Also added a provisional SongInfoDialog (thanks @HitcherUK ) and node files that make the additional filtered artist listings and roles data visible.

Can we merge with this provisonal Dialog? Functionallity it has the thumbs up and is good to go, just a question of the UI aspects. @Razzeee haven't heard from you in a while.

I will get a test build up on the mirrors if anyone would like to try it and see what I am talking about.

DaveTBlake commented:
A little bit more of a twiddle with the song info dialog (thanks @HitcherUK )
jenkins build this please

Tolriq commented:
So as I feared this was pushed without @DaveTBlake reporting any of my concerns so that @Montellese have the correct information to validate ....

Why the changes to GetArtists is wrong : 
1) It adds new filter role and roleid on data that is not part of artists and can not be get by anyone. This is as an API definition quite wrong. (And the fact that there's a wrong precedence does not validate the fact that new bad things should be added !)

2) It adds an hidden undocumented filter on roleid = 1. So now the GetArtists do not return all Artists from the table but only some based on a filter that no one is aware of.

3) To get all artists you need to filter on roleid = -1, first it's not documented, but more wrong : You need to use an undocumented internal magic number to do a filter on a public API. Completely insane.
(And not talking about the fact that doing that filter returns data that you can't really use since you can't apply the filter manually since you do not have access to roledid and role.)

About naming : There's some major incoherence too, it seems it's all over contributor roles, but then it's Library.Fields.Role so not related to audio or contributors and not really clear as it can have many meanings.

"Audio.Contributors" do not include or return roleid only the role string, also not cool for any proper usage of the API. Come on it means API consumers can not recreate a proper joint and need to use 
join on Strings, really on strings ? (Specially when due to current api limitation this joint table is mandatory to recreate correct data for offline use ?)

This is IMO a lot of things that should have been at least discussed with the proper people, and not hidden.

Really disappointed that JSON like that was merged, this open doors to anything in future and is not toward a better base, unlike what some others guys from the team tries to achieve !

Tolriq commented:
Another funny thing : The GetSong part , you can filter on role and roleid (but again roleid not returned in contributor filed).

So basically you can query all songs that have a composer or a lyricist.

If you use 2 filter + "and" you'll have all songs where toto is an artist and there's a composer (but composer could be anyone).

@DaveTBlake when you ask for advice to someone that have the knowledge and despite more than 20 PM / Mails you decide to ignore all the messages about your changes being bad and / or useless you should at least have taken time to ask someone else for confirmation.

@Montellese do not use / care about music so he only checked syntax for JSON part without having all the necessary element to understand all the problems you integrated. And since you were aware of the problems but decide to merge, this is called sabotage.

Sorry to go public about our exchanges but Kodi integrity is important at least to me.

DaveTBlake commented:
@Tolriq I am sorry if I have upset you, but I am surprised and very disappointed in your reaction too and the accusations you make.

I had a long email exchange with Tolriq over the JSON aspects of this PR, and incorporated a number of improvements. However we had some remaining disagreements. Role is much like genre, so I have implemented the API in a similar manner. It turns out that genre and the filters in GetArtists do not work in the way Tolriq thought they did (neither does the core querying of Kodi). He demanded that I treat role differently and said that genre was broken (in fact that whole area of Kodi), and I declined. I feel that the kind of changes needed to get the API to do what wants are beyond the scope of this PR.

As I said to him before I will happily discuss the whole music API design issue in pubic, I will even do any work needed, but I just don't see it belongs in this PR.

Razzeee commented:
It's pre alpha, there is enough time to improve upon this.

On Mon, Jan 18, 2016, 09:22 Tolriq <notifications@github.com> wrote:

> Another funny thing : The GetSong part , you can filter on role and roleid
> (but again roleid not returned in contributor filed).
>
> So basically you can query all songs that have a composer or a lyricist.
>
> If you use 2 filter + "and" you'll have all songs where toto is an artist
> and there's a composer (but composer could be anyone).
>
> @DaveTBlake <https://github.com/DaveTBlake> when you ask for advice to
> someone that have the knowledge and despite more than 20 PM / Mails you
> decide to ignore all the messages about your changes being bad and / or
> useless you should at least have taken time to ask someone else for
> confirmation.
>
> @Montellese <https://github.com/Montellese> do not use / care about music
> so he only checked syntax for JSON part without having all the necessary
> element to understand all the problems you integrated. And since you were
> aware of the problems but decide to merge, this is called sabotage.
>
> Sorry to go public about our exchanges but Kodi integrity is important at
> least to me.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/8015#issuecomment-172459793>.
>

Tolriq commented:
@DaveTBlake you mixed 2 things : The ask for something when you asked me about the API.

And the last mails exchanges where I did warn you about the breaking you introduce and the fact that API was not finished and should not be merged.
You on purpose refused to take time to think about API and force merge. (Removing JSON part and thinking 2 seconds about it would have at least ensure that there's something that can be used...)

Leading to an API that does not even work for the basics (GetSongs that have composers ?) and that introduce breaking to all JSON convention.
Changes to GetArtist are bad (filter on data not on filtered data, implicit filter, magic number to remove a filter) and you were aware of that. This is against the whole JSON handling in Kodi ....

Do not mix with the rest of the discussion.

What you merged does not even provide any way to achieve what you want, offline remote or not....

@Razzeee breaking things on purpose is bad, you know that once things are merged it's complicated, since the API needs change it means a new major bump of JSON just to fix a too quick merge when the author was aware of the problem ?

Tolriq commented:
@DaveTBlake even better : you can filter songs by roleid 50 (all songs are returned) even when there's no roleid 50 existing ;) (And of course filtering by roleid have no purpose since you need to be able to filter by artist + role together to list correct songs)

And you can not use filter roleid -1 for artist as the definition does not allow it and it returns an error.

So 100% of the JSON part is broken and you wonder why I'm upset and react like that ? The JSON part should never have been merged ....
